### PR TITLE
Even more code cleanup

### PR DESCRIPTION
**Pull request by MeuhV1**
Run a 3000 BC and 600 AD Canadian start without errors.

(This will be the last code cleanup for now)

### CODE REVIEW

File: Assets/Python/CvRFCEventHandler.py [hunk 1]
```diff
@@ -129,22 +129,22 @@ def onCityAcquired(self, argsList):
...
 		
 		# relocate capitals
 		if utils.getHumanID() != iPlayer:
-			if iPlayer == iTurkey and (city.getX(), city.getY()) == (68, 45):
-				utils.moveCapital(iTurkey, (68, 45)) # Kostantiniyye
-			elif iPlayer == iMongolia and (city.getX(), city.getY()) == (102, 47):
+			if iPlayer == iTurkey and tCity == (68, 45):
+				utils.moveCapital(iTurkey, tCity) # Kostantiniyye
+			elif iPlayer == iMongolia and tCity == (102, 47):
 				utils.moveCapital(iMongolia, (102, 47)) # Khanbaliq
```
Comment (dguenms) [inline Assets/Python/CvRFCEventHandler.py]: Why not here too?

File: Assets/Python/RFCUtils.py [hunk 2]
```diff
@@ -502,7 +502,7 @@ def relocateGarrisons(self, tCityPlot, iOldOwner):
 			plot = gc.getMap().plot(x, y)
 			iNumUnits = plot.getNumUnits()
 			for i in range(iNumUnits):
-				unit = plot.getUnit(i)
```
Comment (dguenms) [inline Assets/Python/RFCUtils.py:505]: Good catch.

File: Assets/Python/RFCUtils.py [hunk 3]
```diff
@@ -1778,5 +1778,11 @@ def setStateReligionBeforeBirth(self, lPlayers, iReligion):
 				
 	def playerNames(self, lPlayers):
 		return str([gc.getPlayer(iPlayer).getCivilizationShortDescription(0) for iPlayer in lPlayers])
+		
+	def isYearIn(self, iStartYear, iEndYear):
+		iGameTurn = gc.getGame().getGameTurn()
+		if getTurnForYear(iStartYear) <= iGameTurn <= getTurnForYear(iEndYear):
```
Comment (dguenms) [inline Assets/Python/RFCUtils.py]: You could even `return getTurnForYear(iStartYear) <= iGameTurn <= getTurnForYear(iEndYear)`

File: Assets/Python/RiseAndFall.py [hunk 4]
```diff
@@ -478,25 +468,25 @@ def prepareChina(self):
...
 			pBeijing = gc.getMap().plot(tBeijing[0], tBeijing[1])
 			pBeijing.getPlotCity().kill()
 			pBeijing.setImprovementType(-1)
 			pBeijing.setRouteType(-1)
 
 		tCultureRegionTL = (98, 37)
 		tCultureRegionBR = (109, 49)
-		for x in range(tCultureRegionTL[0], tCultureRegionBR[0]+1):       
-			for y in range(tCultureRegionTL[1], tCultureRegionBR[1]+1):     
+		for x in range(tCultureRegionTL[0], tCultureRegionBR[0]+1):
```
Comment (dguenms) [inline Assets/Python/RiseAndFall.py:479]: `for (x, y) in utils.getPlotList(tCultureRegionTL, tCultureRegionBR)`

File: Assets/Python/RiseAndFall.py [hunk 5]
```diff
@@ -523,14 +513,14 @@ def prepareChina(self):
 
 	def setupBirthTurnModifiers(self):
 		for iCiv in range(iNumPlayers):
-			if (iCiv >= iIndia and not gc.getPlayer(iCiv).isHuman()):
+			if iCiv >= iIndia and not gc.getPlayer(iCiv).isHuman():
```
Comment (dguenms) [inline Assets/Python/RiseAndFall.py]: `if tBirth[iCiv] > -3000 and not gc.getPlayer(iCiv).isHuman()`

File: Assets/Python/RiseAndFall.py [hunk 6]
```diff
@@ -523,14 +513,14 @@ def prepareChina(self):
 
 	def setupBirthTurnModifiers(self):
 		for iCiv in range(iNumPlayers):
-			if (iCiv >= iIndia and not gc.getPlayer(iCiv).isHuman()):
+			if iCiv >= iIndia and not gc.getPlayer(iCiv).isHuman():
 				data.players[iCiv].iBirthTurnModifier = gc.getGame().getSorenRandNum(11, "BirthTurnModifier") - 5 # -5 to +5
 		#now make sure that no civs spawn in the same turn and cause a double "new civ" popup
 		for iCiv in range(iNumPlayers):
-			if (iCiv > utils.getHumanID() and iCiv < iBrazil):
+			if iCiv > utils.getHumanID() and iCiv < iBrazil:
```
Comment (dguenms) [inline Assets/Python/RiseAndFall.py]: `if iCiv > utils.getHumanID() and iCiv < iNumPlayers-1`

(Should be Canada by now already)

File: Assets/Python/RiseAndFall.py [hunk 7]
```diff
@@ -1053,235 +1042,124 @@ def checkPlayerTurn(self, iGameTurn, iPlayer):
...
-						    iPlayer == iEthiopia or \
-						    iPlayer == iMali):
-							if (data.iCivsWithNationalism <= 0):
-								iNewCiv = iNative		    
-						splittingCity = cityList[gc.getGame().getSorenRandNum(len(cityList), 'random city')]
-						utils.cultureManager((splittingCity.getX(),splittingCity.getY()), 50, iNewCiv, iPlayer, False, True, True)
-						utils.flipUnitsInCityBefore((splittingCity.getX(),splittingCity.getY()), iNewCiv, iPlayer)			    
-						data.tTempFlippingCity = (splittingCity.getX(), splittingCity.getY())
-						utils.flipCity((splittingCity.getX(),splittingCity.getY()), 0, 0, iNewCiv, [iPlayer])   #by trade because by conquest may raze the city
+						if iRndNum == 1:
```
Comment (dguenms) [inline Assets/Python/RiseAndFall.py:1145]: Not that it matters (I think), but `iRandNum % 2 == 0` and `iRandNum == 1` are opposite cases.

File: Assets/Python/RiseAndFall.py [hunk 8]
```diff
@@ -1955,92 +1717,80 @@ def warOnSpawn(self, iPlayer, lEnemies):
...
 			tSeaPlot = self.findSeaPlots(tCapital, 1, iCiv)
 			
-			if iCiv == iRome:
-				cityList = PyPlayer(iCiv).getCityList()
-				for city in cityList:
-					pCity = city.GetCy()
-					if pCity.getRegionID() == rIberia:
-						tCapital = (pCity.getX(), pCity.getY())
-						break
+			# if iCiv == iRome:
```
Comment (dguenms) [inline Assets/Python/RiseAndFall.py]: Why is this removed?

File: Assets/Python/UniquePowers.py [hunk 9]
```diff
@@ -104,17 +104,14 @@ def vikingUP(self, argsList):
...
 #------------------ROMAN UP-----------------------
 
-	def doRomanWar(self, iCiv):
+	def doRomanWar(self, iCiv): # Unused
 	
 		if iCiv in [iCarthage, iPersia, iCeltia, iEgypt]:
 			iNumTargets = 2
 			self.romanConquestUP(iCiv, iNumTargets)
-		elif iCiv == iGreece and utils.getHumanID() != iGreece:
+		elif iCiv == iRome and utils.getHumanID() != iRome:
```
Comment (dguenms) [inline Assets/Python/UniquePowers.py]: That's still supposed to be Greece: iCiv is the target.

File: Assets/Python/UniquePowers.py [hunk 10]
```diff
@@ -233,17 +206,17 @@ def tamilConquestUP(self, iEnemy, iNumTargets=1):
 
 #------------------ARABIAN U.P.-------------------
 
-	def mughalUP(self, city):
+	def mughalUP(self, city): # Unused
 		return
-		pMughals = gc.getPlayer(iMughals)
+		# pMughals = gc.getPlayer(iMughals)
```
Comment (dguenms) [inline Assets/Python/UniquePowers.py:211]: pMughals is used right in the next line, that's is going to fail (if the method is ever used again).

File: Assets/Python/UniquePowers.py [hunk 11]
```diff
@@ -233,17 +206,17 @@ def tamilConquestUP(self, iEnemy, iNumTargets=1):
...
 		iStateReligion = pMughals.getStateReligion()
 
 		if iStateReligion >= 0:
 			city.setHasReligion(iStateReligion, True, True, False)
 
-	def seljukUP(self, city):
+	def seljukUP(self, city): # Unused
 		return
-		pSeljuks = gc.getPlayer(iSeljuks)
+		# pSeljuks = gc.getPlayer(iSeljuks)
```
Comment (dguenms) [inline Assets/Python/UniquePowers.py]: See above.

File: Assets/Python/UniquePowers.py [hunk 12]
```diff
@@ -260,71 +233,61 @@ def seljukUP(self, city):
 		
 
 	def arabianUP(self, city):
-		pArabia = gc.getPlayer(iArabia)
+		#pArabia = gc.getPlayer(iArabia)
```
Comment (dguenms) [inline Assets/Python/UniquePowers.py]: See above.

File: Assets/Python/UniquePowers.py [hunk 13]
```diff
@@ -260,71 +233,61 @@ def seljukUP(self, city):
...
+		if pWinningPlayer.getID() != iAztecs:
 			return
 			
-		if utils.isReborn(iAztecs): return
-
 		pLosingPlayer = gc.getPlayer(pLosingUnit.getOwner())
 		cLosingUnit = PyHelpers.PyInfo.UnitInfo(pLosingUnit.getUnitType())
 		
-		if (pLosingUnit.getUnitType() < iWarrior):
+		if pLosingUnit.getUnitType() < iWarrior:
```
Comment (dguenms) [inline Assets/Python/UniquePowers.py]: I think this is supposed to exclude animals, so it'd be better to use `pLosingUnit.isAnimal()`.

File: Assets/Python/Victory.py [hunk 14]
```diff
@@ -2626,13 +2622,8 @@ def getTotalCulture(lPlayers):
...
 def countImprovements(iPlayer, iImprovement):
-	iCount = 0
-	for x in range(iWorldX):
-		for y in range(iWorldY):
-			plot = gc.getMap().plot(x, y)
-			if plot.getOwner() == iPlayer and plot.getImprovementType() == iImprovement:
-				iCount += 1
-	return iCount
+	if iImprovement <= 0: return 0
+	return gc.getPlayer(iPlayer).getImprovementCount(iImprovement)
```
Comment (dguenms) [inline Assets/Python/Victory.py:2626]: Nice!

**Review by dguenms — CHANGES_REQUESTED** [review]
Only a few minor requests / questions. I've also added some suggestions for things I noticed along the way, if you don't want to cover them just let me know and I'll do it myself.

Will also run a couple of test starts for myself.

Thank you for doing this, the code looks much better now, and you've covered a lot of improvements I wanted to do later myself.

File: Assets/Python/RiseAndFall.py (hunk 4 above)
Comment (MeuhV1) [reply Assets/Python/RiseAndFall.py:479]: I did do this, but somehow, I got an Unidentifiably C++ exception @ line 493,pCurrent.setOwner(-1).

File: Assets/Python/RiseAndFall.py (hunk 7 above)
Comment (MeuhV1) [reply Assets/Python/RiseAndFall.py:1145]: True. I just wanted to get rid of the modulo calculation, as it is not required.

File: Assets/Python/RiseAndFall.py (hunk 8 above)
Comment (MeuhV1) [reply Assets/Python/RiseAndFall.py]: The only thing that changes is tCapital. But tCapital is not used anymore in that function. I assume it should be before tSeaPlot = self.findSeaPlots(tCapital, 1, iCiv), so it is used.

File: Assets/Python/UniquePowers.py (hunk 10 above)
Comment (MeuhV1) [reply Assets/Python/UniquePowers.py:211]: I doubt that. The pMughals is also initiated in the Consts. So it should be loaded as well.

File: Assets/Python/UniquePowers.py (hunk 13 above)
Comment (MeuhV1) [reply Assets/Python/UniquePowers.py]: Also settlers, workers, scouts, explorers and missionaries.

pLosingUnit.isAnimal() or pLosingUnit.getDomainType() == DomainTypes.DOMAIN_LAND and pLosingUnit.getCombat() > 0
(This will include scouts and explorers to be enslaved, but I don't think that is a problem)

File: Assets/Python/RiseAndFall.py (hunk 4 above)
Comment (dguenms) [reply Assets/Python/RiseAndFall.py:479]: Okay, let's leave it like this for now.

File: Assets/Python/RiseAndFall.py (hunk 7 above)
Comment (dguenms) [reply Assets/Python/RiseAndFall.py:1145]: Yeah that's good.

File: Assets/Python/UniquePowers.py (hunk 10 above)
Comment (dguenms) [reply Assets/Python/UniquePowers.py:211]: Oh right, forgot about that.

File: Assets/Python/UniquePowers.py (hunk 13 above)
Comment (dguenms) [reply Assets/Python/UniquePowers.py]: Agreed, that's better.

File: Assets/Python/RiseAndFall.py (hunk 8 above)
Comment (dguenms) [reply Assets/Python/RiseAndFall.py]: Right, moving it would be better.

**Comment by dguenms** [comment]
Also didn't run into any errors in my autoplay tests.

**Review by dguenms — APPROVED** [review]
